src/main/java/com/commercetools/sync/cartdiscounts/CartDiscountSync.java

```diff
@@ -202,12 +204,21 @@ private void handleError(@Nonnull final String errorMessage,
                         .resolveReferences(newCartDiscount)
                         .thenCompose(resolvedDraft -> syncDraft(oldCartDiscountMap, resolvedDraft))
                         .exceptionally(completionException -> {
-                            final ReferenceResolutionException referenceResolutionException =
-                                (ReferenceResolutionException) completionException.getCause();
-                            final String errorMessage = format(FAILED_TO_RESOLVE_REFERENCES, newCartDiscount.getKey(),
-                                             referenceResolutionException.getMessage());
-                            handleError(errorMessage, referenceResolutionException, 1);
-                            return null;
+                            if (completionException.getCause() instanceof ReferenceResolutionException) {
```

Comment:
why do you need getCause in here? checking the expectation itself was not enough?

Reply:
java.util.concurrent.CompletionException is thrown, and the actual cause is wrapped inside.

Reply:
what if it's not caused with that way, we would not know it before testing it with different exception classes with different cause types, I also think that it would work without a problem, but a code should always be running against an automation test rather than expecting that it would work without a problem.

Reply:
It is behaviour of CompletableFuture class. The wrapping occurs in the following line
https://github.com/commercetools/commercetools-sync-java/blob/7baae24007cbb49afe7a6345bfee535765f3ac7b/src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java#L116

imo if we don't return the exception in such way, we have to change it as try-catch pattern simply in order to get rid of completionException.getCause(). Any other suggestion?

Reply:
As discussed with Roman, the thrown exception will be passed into handleError directly and let user to deal with it. If-else checking is no longer used in Sync classes.